sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -157,6 +157,14 @@ private[hive] class HiveMetastoreCatalog(hive: HiveContext) extends Catalog with
         tbl.setProperty(s"spark.sql.sources.schema.part.${index}", part)
       }
     }
+
+    // Set the fields for consistent with Hive MetaStore API(.getCols()) accessing
+    val schema = ResolvedDataSource(hive, userSpecifiedSchema, provider, options).relation.schema
+    tbl.setFields(schema.iterator.map { attr =>
+      new FieldSchema(attr.name, toMetastoreType(attr.dataType), null)
```

Comment:
We can't limit ourselves to the lowest common denominator.  I would be okay with lying about the schema in this case and printing a warning or something that it might be incompatibly with other systems.